src/test/test-transformers.ts

```diff
@@ -29,66 +32,135 @@ describe("transformers", () => {
+					"Not all objects were successfully processed"
+				);
 			}
+		}
+
+		describe("docs example", () => {
+			it(
+				"passes part 1 of example from docs",
+				clockwise(async () => {
+					expectErrorFromEnd = true;
+					// Without endCatcher()
+					const resultPromise = getRowsFromDatabase<MyObject>(myQuery)
+						.transform(mapper(identity))
+						.forEach(process, () => {
+							if (someCondition) {
+								throw new NotAllObjectsProcessedError();
+							}
+						});
 
-			(() => {
-				// Without endCatcher()
-				const source = new Stream<string>();
-				source.write("Craig");
-				source.write("Jolene");
-				source.write("Sam");
-				source.write("Cassandra");
-				source.write("Xavier");
-				source.end().catch((e) => {
-					if (e.message === "Too many guests!") {
-						requestBiggerTable();
-					}
-				});
-
-				expect(
-					source.forEach(addGuestToReservation, () => {
-						if (guestsAdded() > 4) {
-							throw new Error("Too many guests!");
-						}
-					})
-				).rejectedWith(Error); // This throws an unhandled promise rejection error, even though the source handled it!
-			})();
-
-			(() => {
-				// With endCatcher()
-				const source = new Stream<string>();
-				source.write("Craig");
-				source.write("Jolene");
-				source.write("Sam");
-				source.write("Cassandra");
-				source.write("Xavier");
-				source.end();
-
-				source
+					await expect(
+						resultPromise
+					).eventually.rejected.and.be.instanceOf(
+						NotAllObjectsProcessedError
+					); // This throws an unhandled promise rejection error, even if the source handles it!
+				})
+			);
+
+			it(
+				"passes part 2 of example from docs",
+				clockwise(async () => {
+					expectErrorFromEnd = false;
+					// With endCatcher()
+					const resultPromise = getRowsFromDatabase<MyObject>(myQuery)
+						.transform(
+							endCatcher((err: Error) => {
+								if (
+									err instanceof NotAllObjectsProcessedError
+								) {
+									// ignore
+								} else {
+									// re-throw all other errors
+									throw err;
+								}
+							})
+						)
+						.transform(mapper(identity))
+						.forEach(noop, () => {
+							if (someCondition) {
+								throw new NotAllObjectsProcessedError();
+							}
+						});
+
+					await expect(resultPromise).eventually.equals(undefined); // No longer throws the error, because handleError() returns successfully
+				})
+			);
+		});
+
+		it(
+			"Indeed throws an error if endCatcher() throws",
+			clockwise(async () => {
+				function getSource() {
+					const source = new Stream<number>();
+					source.write(1);
+					source.write(2);
+					expect(source.end()).to.eventually.rejectedWith(Error);
+					return source;
+				}
+
+				const resultPromise = getSource()
 					.transform(
-						endCatcher((e) => {
-							if (e.message === "Too many guests!") {
-								requestBiggerTable();
+						endCatcher((err: Error) => {
+							if (err.message !== "This is fine actually") {
```

Comment:
I was trying to follow what you're doing, only to find that you basically always want to `throw err` here. Just remove the `if`?

Reply:
Felt silly to have an implementation of endCatcher that was effectively a `noop`, but, yes, it is clearer to do it that way.